Outline the `recipe` specification

therazix opened the pull request:
Introduce draft of tmt recipe specification as the first step towards implementing support for executing a subset of tests based on reported results.

The recipe captures a complete, static snapshot of a tmt run and can be reused to reproduce the run with the exact same configuration, without requiring additional input. It contains preprocessed information about plans, tests, and run configuration, eliminating the need for dynamic evaluation.


Resolves: #3854
Related: [TT-273](https://issues.redhat.com/browse/TT-273)

Pull Request Checklist

* [x] update the specification
* [x] include a release note

psss commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,87 @@
+        # Run  configuration
+        run:
+            root: "/path/to/fmf/root"
+            remove: false
```
In order to be able to reproduce the run we need the git repo details as well. It should be basically fmf id, just without the name, so `url` of the repo, `ref` in the form of a hash and `path` to  the metadata tree. The `name` part should be covered by plan names under the `plans` key below.

Also, I guess, we should store here the original `environment` and `context` with which the run was created? The complete `environment` should be stored for each test in `tests.yaml`, but the environment variables are also delivered to other places, such as scripts executed during the `prepare` step.

psss commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,87 @@
+                    results:
+                      - name: /test/name
+                        result: pass
+                        ...
```
Shall we explicitly mention that when submitting a new run based on a recipe the `results` key is ignored and new results will be created based on the actual test execution? Or do we expect some integration of the past results as well?

psss commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,87 @@
+                    tests:
+                      - name: /test/name
+                        summary: Test summary
+                        ...
```
In order to be able to execute tests the expanded test metadata will not be enough. Although we do not expect any `adjust` rules to be evaluated or any other dynamic steps, we will have to fetch the git repositories again so that the test code itself is available. Perhaps a short mention of this should be added here or in the story description?

psss reviewed (approved):
Thanks for addressing all comments. Looks good now. I think the proposed outline is a sufficient content to kick off the discussion. Just added a release note and the related issue for reference in aedbc42f.

happz commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,109 @@
+    Generated recipe can be edited, manually or by using a script, which allows users
+    to select only a subset of discovered tests to be reexecuted.
+
+    .. code-block:: yaml
```
I would strongly recommend following the type/description documented format from Results and HW requirement specification:

```yaml
.. code-block:: yaml

    # Mapping, stores git repository configuration
    git:
      # String, URL of the repository.
      url: http://some.git.host.com/project/tests.git
      # String, git reference - branch, commit, tag - to check out in the repository.
      ref: main
      # String, path to the fmf root in the repository.
      path: /some/path
```

And so on.

happz commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,109 @@
+
+    .. code-block:: yaml
+
+        # Git repo configuration
```
Which repository? There are many repositories, e.g. `discover` phases can clone many. Let's be specific.

happz replied:
Maybe don't even call it `git`, can we find a better, more specific name, e.g. `something-repository`?

happz replied:
`run-repository`?

happz commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,109 @@
+                        how: shell
+                        order: 50
+                        ...
+
```
`cleanup` deserves a spot too.

therazix replied:
Added in 2455271b796ee94ccd8e449277e0e624f289cfdc.

happz commented:
> * Since the recipe should be static and already evaluated, the tests specified in the recipe will already have all the necessary environment, including variables inherited from the plan and the command line. However, we also need to keep the plan and command-line environment, because some steps (for example the prepare step) might need them. If the user modifies the plan environment, should it also be propagated to the tests? I can see this could cause confusion.

You should try to avoid repetition: if you need standalone plan environment, for whatever reason, then it shouldn't be baked into test environments. If I modify plan-level environment, do I need to modify also all the tests? I think we will need to save several components of the environment, and compose it in run time.

> * The git key: As far as I know, this key would be used by Testing Farm to clone the git repository containing the tests, before tmt is even executed. Should the recipe include this key at all? tmt doesn't handle the cloning itself, and the use case seems to be limited to TF only. For local use, this key has no effect. Should we even include it?

Hard to tell, maybe not. It's even possible to not have git repo at all. As you said, TF does the cloning, maybe this responsibility lies with the caller.

> * How should test requirements be handled? Tests may require additional files. Where should we save them? A similar issue is with Ansible playbooks, but these could potentially be stored directly inside the recipe.

Tricky, but I would prefer to not save anything in the recipe unless absolutely necessary. `prepare/shell` is getting `url` option, and we are not going to save its content - we will save the URL, precise ref, so we can repeat the clone exactly. Maybe we should do the same with `ansible` - drop the "remote playbook specified by URL" feature, which is complicated because it's just the URL, and instead adopt what `shell` does, actual git repo whose properties we can save and repeat.

WRT test requirements: they are either in the same repo, then we have them, or they are from some other place, if it's a git repo, we can save its properties again. If it's not, we need to find a way to make its acquisition repeatable.

happz commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,109 @@
+                      - name: default-0
+                        how: ansible
+                        order: 50
+                        # Playbooks are included directly in the recipe
```
See my comment below:

> Tricky, but I would prefer to not save anything in the recipe unless absolutely necessary. prepare/shell is getting url option, and we are not going to save its content - we will save the URL, precise ref, so we can repeat the clone exactly. Maybe we should do the same with ansible - drop the "remote playbook specified by URL" feature, which is complicated because it's just the URL, and instead adopt what shell does, actual git repo whose properties we can save and repeat.

So I would throw the whole "remote, git-less playbook" feature out of the window, and I'd replace it with git url/ref.

psss commented:
Too late for 1.54, moving.

mcasquer reviewed (approved):
LGTM

kkaarreell commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,109 @@
+                        order: 50
+                        ...
+                    # Test results
+                    results:
```
Working with two sources would make such post-processing more complicated, I am afraid.

kkaarreell replied:
Results are per plan, right? Imagine someone rerunning his request expanded into 10+ plans trying to utilize this feature for rerunning failed tests.

thrix replied:
thank you, resolving

kkaarreell commented:
Does the current structure of `results` play well with multi-host tests? Just checking...

thrix commented:
We also discussed the possible flavor for the recipes. Because currently the main use cases for recipes are just prescriptions to rerun tests, storing hashes of remote repositories and `ansible-playbooks` is not needed. Seems we do not have currently use cases for "strict" reproducibility for now.

therazix commented:
> We also discussed the possible flavor for the recipes. Because currently the main use cases for recipes are just prescriptions to rerun tests, storing hashes of remote repositories and `ansible-playbooks` is not needed. Seems we do not have currently use cases for "strict" reproducibility for now.

Removed the mention of commit hashes and changed the reference in 4f125d3a8acf35dbdf7d7331bb173c60861d9c5d.

happz commented:
> I also edited the description of run-repository in [5978ceb](https://github.com/teemtee/tmt/commit/5978cebc165bb151ca37c851f10f6f28b55376b5), but I’m still wondering if we should include it at all.

I would drop it: we can always add it later, and we did not make up our minds yet about what user should be responsible for. We can always add it if needed. Once we add it and realise it's not a good idea, it would be harder to drop it.

therazix commented:
> > I also edited the description of run-repository in [5978ceb](https://github.com/teemtee/tmt/commit/5978cebc165bb151ca37c851f10f6f28b55376b5), but I’m still wondering if we should include it at all.
> 
> I would drop it: we can always add it later, and we did not make up our minds yet about what user should be responsible for. We can always add it if needed. Once we add it and realise it's not a good idea, it would be harder to drop it.

Alright, dropped in 9c2d4b29ebe25bf579a31f0be3b1e497a37f8a1d.

psss commented on spec/recipe.fmf:
```diff
@@ -0,0 +1,116 @@
+        recipe execution.
+
+    .. note::
+        When submitting a new run based on a recipe, the ``results`` key of the
```
```suggestion
        When submitting a new run based on a recipe, the ``results-path`` key of the
```

Here it should be updated as well.

psss reviewed (approved):
Looks good to me, just three minor suggestions.
